### ISSUE TITLE
仿照Doom-emacs的方式优化配置?

### ISSUE THREAD

**Issue by cireu** [seagle0128/.emacs.d#37]
[Doom-emacs](https://www.github.com/hlissner/doom-emacs)的作者提到启动Emacs时应当最小化`load-path`, 避免Emacs到每个`load-path`下面寻找文件造成性能损失. 我觉得可以用类似陈斌的方法进行优化

```elisp

;; No need now
;; (add-to-list 'load-path (expand-file-name "lisp" user-emacs-directory))

(defmacro require-init (config)
  "Load your init files"
  `(load ,(file-truename (format "~/.emacs.d/lisp/%s) config)))

(require-init init-const)
(require-init init-custom)
;; Blabla...
```
我个人在windows上体验了一下, emacs启动时间由16.5s直接变为11.8s, 瞬间泪流满面....
而且这种改动对配置文件的影响也会比较小, 没有doom-emacs那种极端且复杂的优化,
不会破坏配置文件结构的完整性. 我觉得可以考虑一下

**Comment by seagle0128** [comment]
谢谢你的建议！
因为我大多在 Linux 和 macOS 上开发，启动时间已经控制在2s 以内，没有主要到会有这么明显的差别。Windows 确实启动非常慢，所以我一般都打开不会关闭。这个方案我需要测试一段看看效果，评估下值不值得改动。

**Comment by seagle0128** [comment]
经过测试，这种方法缺水有效，大概能缩短~20%的启动时间。测试环境：
```
macOS Majove, GNU Emacs 26.1
Ubuntu 18.04，GNU Emacs 27.0.50.
```
究其原因，应该是package 包初始化之后会加载所有的安装包路径，而`lisp`目录在相对靠后的位置，这样每次加载配置文件会花时间找路径。这也让我想到了另外一种解决方案，就是把`lisp`的路径放到最前面，理论上也会缩短加载的时间，而且改动更小。我在 mscOS 和 Linux 上测试和chenbin 的方法差不多，但不太清楚 Windows 上的情况。

你能帮忙在 Windows 上对比下两种方法吗？把上述代码换成下面即可，调用的地方仍然用`require`。 

```elisp
;; Load path
(add-to-list 'load-path (expand-file-name "lisp" user-emacs-directory))

(defadvice package-initialize (after my-init-load-path activate)
  "Reset `load-path'."
  (cl-pushnew (expand-file-name "lisp" user-emacs-directory) load-path)
  (cl-pushnew (expand-file-name "site-lisp" user-emacs-directory) load-path))
```

**Comment by cireu** [comment]
wow, 似乎你的方案比红还要快!

测试环境: Windows 10 1803 Emacs 26.1

| | 宏 | 直接添加`load-path` | 新优化方案 |
| `emacs-init-time` | 8.8s | 12.9s | 8.1s |
| `benchmark-init` | 23241ms | 26161ms | 20793ms |

这可以说是相当惊喜了, 我猜测宏版本比较慢的原因是我没有进行字节码编译 (一般我也不会直接编译 `init.el` , 不然实在是太不方便了. ), 反复调用了`file-truename`这个函数吃了亏. 我这个宏还是直接从doom哪里提取出来的而不是直接抄的陈斌的, 陈斌的那个就算字节码编译完也不会从这里有优化, 他的comma点错了地方. 

无论如何, 谢谢你所做的工作! 另外佩服你的机智 😄

**Comment by seagle0128** [comment]
Cool!!! Windows 版本 Emacs 读取文件效率确实不好，毕竟是 msys 编译的。我刚才也简单测试了下，我的Windows环境里大概能缩短~30%的时间。 看起来这是个有效且简洁的solution。也非常感谢你提出的建议，才能想到这样的解决方案。 Thanks！ 😄

**Comment by seagle0128** [comment]
重构了下，有时间请帮忙测试下。注意，这里只能使用`push`而不是`cl-pushnew`.

```elisp
(defun update-load-path (&rest _)
  "Update `load-path'."
  (push (expand-file-name "site-lisp" user-emacs-directory) load-path)
  (push (expand-file-name "lisp" user-emacs-directory) load-path))

;; Optimize: Force "lisp"" and "site-lisp" at the head to reduce the startup time.
(advice-add #'package-initialize :after #'update-load-path)

(update-load-path)
```

**Comment by redguardtoo** [comment]
We I need advice? could I just call `update-load-path` before `package-initialize`?

**Comment by cireu** [comment]
谢谢, 两者基本上没有太大的区别. 不用`cl-pushnew`的原因是`push`的代码会比较简单一些吗?

我之前从未注意到`file-truename`的性能问题, 被你提醒后发现它是一个elisp函数, 而`expand-file-name`是C函数. 看来宏的问题的确是出在`file-truename`这里了. 

另外, 我发现如果用`use-pacakge`指定`:load-path`关键字, `use-package`会自作聪明的也把那个path自己`push`到`load-path`的顶端. 可能用宏比较适合我这种用其他包比较多的人, Thanks!

**Comment by cireu** [comment]
> 
> 
> Why advice? could I just call `update-load-path` before `package-initialize`?

`update-load-path` needs to be called after `package-initialize`. However, `package-initialize` was fired up in `init-package.el` in his config. So he need Emacs can require `init-package.el` first.

**Comment by seagle0128** [comment]
@redguardtoo @cireu has explained above. I need to load some config files in `~/.emacs.d/lisp/` first, and keep lisp directory at the head of load-path to speedup after `package-iniitialize`.

**Comment by redguardtoo** [comment]
place `(package-initalize)` at the begining of `init.el`,  `update-load-path` is placed at next line, then you don't need advice.

**Comment by seagle0128** [comment]
@redguardtoo Right! But in 27, `package-initalize` call is unnecessary, I am not very sure the new behaviors. `advice` is the safer way, but I am thinking to remove it after the testing on 27.

**Comment by redguardtoo** [comment]
Good point.
